File: appkillermanager/src/main/java/com/thelittlefireman/appkillermanager/managers/KillerManager.java
package com.thelittlefireman.appkillermanager.managers;

import android.content.Context;
import android.content.Intent;

import androidx.annotation.Nullable;

import com.thelittlefireman.appkillermanager.devices.DeviceBase;
import com.thelittlefireman.appkillermanager.utils.ActionsUtils;
import com.thelittlefireman.appkillermanager.utils.LogUtils;
import com.thelittlefireman.appkillermanager.utils.SystemUtils;

public class KillerManager {

    public static DeviceBase getDevice() {
        return DevicesManager.getDevice();
    }

    public static boolean isActionAvailable(Context context, Actions actions) {
        DeviceBase sDevice = DevicesManager.getDevice();
        boolean actionAvailable = false;
        if (sDevice != null) {
            switch (actions) {
                case ACTION_AUTOSTART:
                    actionAvailable = sDevice.isActionAutoStartAvailable(context);
                    break;
                case ACTION_POWERSAVING:
                    actionAvailable = sDevice.isActionPowerSavingAvailable(context);
                    break;
                case ACTION_NOTIFICATIONS:
                    actionAvailable = sDevice.isActionNotificationAvailable(context);
                    break;
            }
        }
        return actionAvailable;
    }

    //private static DeviceBase sDevice;

    public static boolean isAnyActionAvailable(Context context) {
        for (Actions action : Actions.values()) {
            if (isActionAvailable(context, action)) {
                return true;
            }
        }
        return false;
    }

    public static boolean isDeviceSupported() {
        return DevicesManager.getDevice() != null;
    }

    /**
     * Return the intent for a specific action
     *
     * @param context the current context
     * @param actions the wanted actions
     * @return the intent
     */
    @Nullable
    private static Intent getIntentFromAction(Context context, Actions actions) {
        DeviceBase sDevice = DevicesManager.getDevice();
        if (sDevice != null) {
            Intent intent = null;
            switch (actions) {
                case ACTION_AUTOSTART:
                    intent = sDevice.getActionAutoStart(context);
                    break;
                case ACTION_POWERSAVING:
                    intent = sDevice.getActionPowerSaving(context);
                    break;
                case ACTION_NOTIFICATIONS:
                    intent = sDevice.getActionNotification(context);
                    break;
            }
            if (intent != null && ActionsUtils.isIntentAvailable(context, intent)) {
                // Intent found action succeed
                return intent;
            } else {
                LogUtils.e(KillerManager.class.getName(), "INTENT NOT FOUND :" +
                        ActionsUtils.getExtrasDebugInformations(intent) + "Actions \n" +
                        actions.name() + "SYSTEM UTILS \n" +
                        SystemUtils.getDefaultDebugInformation() + "DEVICE \n" +
                        sDevice.getExtraDebugInformations(context));
                // Intent not found action failed
                return null;
            }
        } else {
            // device not found action failed
            return null;
               /* LogUtils.e(KillerManager.class.getName(), "DEVICE NOT FOUND" + "SYSTEM UTILS \n" +
                        SystemUtils.getDefaultDebugInformation());*/
        }
    }

    /**
     * Execute the action
     *
     * @param context the current context
     * @param actions the wanted action to execute
     */
    private static void doAction(Context context, Actions actions) throws NoActionFoundException {
        // Avoid main app to crash when intent denied by using try catch
        try {
            Intent intent = getIntentFromAction(context, actions);
            if (intent != null && ActionsUtils.isIntentAvailable(context, intent)) {
                context.startActivity(intent);
                // Intent found action succeed
            }

        } catch (Exception e) {
            // Exception handle action failed
            LogUtils.e(KillerManager.class.getName(), e.getMessage());
            throw new NoActionFoundException();
        }
    }

    /**
     * Execute the action
     *
     * @param activity the current activity
     * @param actions  the wanted action to execute
     *//*
    private static void doAction(Activity activity, Actions actions, Integer code) throws NoActionFoundException {
        // Avoid main app to crash when intent denied by using try catch
        try {
            Intent intent = getIntentFromAction(activity, actions);
            if (intent != null && ActionsUtils.isIntentAvailable(activity, intent)) {
                if (code == null) {
                    activity.startActivity(intent);
                    // Intent found action succeed
                }else {
                    activity.startActivityForResult(intent, code);
                }

            }

        } catch (Exception e) {
            // Exception handle action failed
            LogUtils.e(KillerManager.class.getName(), e.getMessage());
            throw new NoActionFoundException();
        }
    }*/
    public static void doActionAutoStart(Context context) throws NoActionFoundException {
        doAction(context, Actions.ACTION_AUTOSTART);
    }

    public static void doActionNotification(Context context) throws NoActionFoundException {
        doAction(context, Actions.ACTION_NOTIFICATIONS);
    }
    /*
     */

    public static void doActionPowerSaving(Context context) throws NoActionFoundException {
        doAction(context, Actions.ACTION_POWERSAVING);
    }

    public enum Actions {
        ACTION_AUTOSTART("ACTION_AUTOSTART"),
        ACTION_NOTIFICATIONS("ACTION_NOTIFICATIONS"),
        ACTION_POWERSAVING("ACTION_POWERSAVING");

        private String mValue;

        Actions(String value) {
            this.mValue = value;
        }

        public String toString() {
            return this.mValue;
        }
    }

    public static class NoActionFoundException extends Exception {
        NoActionFoundException() {
            this("Intent couldn't find action");
        }

        NoActionFoundException(String message) {
            super(message);
        }
    }

    /*public static void doActionAutoStart(Activity activity, Integer code) throws NoActionFoundException {
        doAction(activity, Actions.ACTION_AUTOSTART, code );
    }

    public static void doActionNotification(Activity activity, Integer code) throws NoActionFoundException {
        doAction(activity, Actions.ACTION_NOTIFICATIONS, code);
    }

    public static void doActionPowerSaving(Activity activity, Integer code) throws NoActionFoundException {
        doAction(activity, Actions.ACTION_POWERSAVING, code);
    }*/
}

File: appkillermanager/src/main/java/com/thelittlefireman/appkillermanager/managers/KillerManager.kt
package com.thelittlefireman.appkillermanager.managers

import android.content.Context
import android.content.Intent
import com.thelittlefireman.appkillermanager.devices.Device
import com.thelittlefireman.appkillermanager.utils.ActionsUtils
import com.thelittlefireman.appkillermanager.utils.SystemUtils
import timber.log.Timber

object KillerManager {

    enum class Action {
        ACTION_AUTO_START,
        ACTION_POWER_SAVING,
        ACTION_NOTIFICATIONS
    }

    @JvmStatic
    val device: Device?
        get() = DevicesManager.getDevice()

    @JvmStatic
    val isDeviceSupported: Boolean
        get() = DevicesManager.getDevice() != null

    @JvmStatic
    fun isActionAvailable(context: Context, action: Action?) = DevicesManager.getDevice()?.let {
        when (action) {
            Action.ACTION_AUTO_START -> it.isActionAutoStartAvailable(context)
            Action.ACTION_POWER_SAVING -> it.isActionPowerSavingAvailable(context)
            Action.ACTION_NOTIFICATIONS -> it.isActionNotificationAvailable(context)
            else -> false
        }
    } ?: false

    @JvmStatic
    fun isAnyActionAvailable(context: Context) = Action.values().any { isActionAvailable(context, it) }

    @JvmStatic
    @Throws(NoActionFoundException::class)
    fun doActionAutoStart(context: Context) {
        doAction(context, Action.ACTION_AUTO_START)
    }

    @JvmStatic
    @Throws(NoActionFoundException::class)
    fun doActionNotification(context: Context) {
        doAction(context, Action.ACTION_NOTIFICATIONS)
    }

    @JvmStatic
    @Throws(NoActionFoundException::class)
    fun doActionPowerSaving(context: Context) {
        doAction(context, Action.ACTION_POWER_SAVING)
    }

    /**
     * Execute the action
     *
     * @param context the current context
     * @param action the wanted action to execute
     */
    @Throws(NoActionFoundException::class)
    private fun doAction(context: Context, action: Action) {
        // Avoid main app to crash when intent denied by using try catch
        try {
            val intent = getIntentFromAction(context, action)
            if (intent != null && ActionsUtils.isIntentAvailable(context, intent)) {
                context.startActivity(intent)
                // Intent found action succeed
            }
        } catch (e: Exception) {
            // Exception handle action failed
            Timber.e(e)
            throw NoActionFoundException()
        }
    }

    /**
     * Return the intent for a specific action
     *
     * @param context the current context
     * @param action the wanted actions
     * @return the intent
     */
    private fun getIntentFromAction(context: Context, action: Action): Intent? {
        val sDevice = DevicesManager.getDevice()
        return if (sDevice != null) {
            val intent = when (action) {
                Action.ACTION_AUTO_START -> sDevice.getActionAutoStart(context)
                Action.ACTION_POWER_SAVING -> sDevice.getActionPowerSaving(context)
                Action.ACTION_NOTIFICATIONS -> sDevice.getActionNotification(context)
            }
            if (intent != null && ActionsUtils.isIntentAvailable(context, intent)) {
                // Intent found action succeed
                intent
            } else {
                Timber.e("""
                    INTENT NOT FOUND :${ActionsUtils.getExtrasDebugInformations(intent)}
                    Actions: ${action.name}
                    SYSTEM UTILS: ${SystemUtils.getDefaultDebugInformation()}
                    DEVICE: ${sDevice.getExtraDebugInfo(context)}
                    """.trimIndent()
                )
                // Intent not found action failed
                null
            }
        } else {
            // device not found action failed
            Timber.w("DEVICE NOT FOUND. %s", SystemUtils.getDefaultDebugInformation())
            null
        }
    }
}
